### PR TITLE
dist type definitions

### PR DESCRIPTION
**Pull request by rosendi**
In this angularconnect's video (https://www.youtube.com/watch?v=_TDUV9R09PM) at 15:00 Bill Ticehurst recommends to start distributing the dts files with a package itself.

### CODE REVIEW

**Comment by guybedford** [comment]
Thanks, sure I'm open to this, but we would need to provide this a little more comprehensively I think if it has to be maintained here. There is an API document at https://github.com/systemjs/systemjs/blob/master/docs/system-api.md which covers everything that should be mentioned.

File: typings/systemjs.d.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+interface System {
+  import(name: string): any;
+  defined: any;
+  amdDefine: () => void;
+  amdRequire: () => void;
+  baseURL: string;
+  paths: { [key: string]: string };
+  meta: { [key: string]: Object };
```
Comment (guybedford) [inline typings/systemjs.d.ts]: Are these necessary? These are actually semi-private properties.

Comment (rosendi) [inline typings/systemjs.d.ts]: This type definitions from the TSD package manager, however now I see that it should be updated. I'll do it tomorrow.

File: typings/systemjs.d.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    * @returns A Promise for the completion of which ever callback is executed.
+    */
+    then<TResult>(onfulfilled?: (value: T) => TResult | PromiseLike<TResult>, onrejected?: (reason: any) => TResult | PromiseLike<TResult>): PromiseLike<TResult>;
+    then<TResult>(onfulfilled?: (value: T) => TResult | PromiseLike<TResult>, onrejected?: (reason: any) => void): PromiseLike<TResult>;
+}
+
+/**
+ * Represents the completion of an asynchronous operation
+ */
+interface Promise<T> {
```
Comment (rosendi) [inline typings/systemjs.d.ts]: Once it is ready, I remove Promise interface and rebase the commits.

File: typings/systemjs.d.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+  modules?: {
+    [key: string]: SystemMetaConfig
+  }
+}
+
+interface SystemConfig {
+  baseURL?: string;
+
+  babelOptions?: any;
```
Comment (rosendi) [inline typings/systemjs.d.ts]: Should we declare `babelOptions`, `traceurOptions`, `typescriptOptions`?

File: typings/systemjs.d.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+  /**
+   * Deletes a module from the registry by normalized name.
+   */
+  delete(moduleName: string): boolean;
+
+  /**
+   * Returns a module from the registry by normalized name.
+   */
+  get(moduleName: string): Module;
```
Comment (rosendi) [inline typings/systemjs.d.ts]: Could you help me with `Module`? Looks like `module` could be generic. How can we describe `Module`?

File: typings/systemjs.d.ts (hunk 3 above)
Comment (rosendi) [inline typings/systemjs.d.ts]: Should we clarify babelOptions, traceurOptions, typescriptOptions?

Comment (guybedford) [inline typings/systemjs.d.ts]: I would leave these out entirely, as they are somewhat custom properties.

Comment (guybedford) [inline typings/systemjs.d.ts]: I guess any is fine for now too though.

File: typings/systemjs.d.ts (hunk 4 above)
Comment (guybedford) [inline typings/systemjs.d.ts]: You will know better than I do. The only thing we can really assume is that it is an object. Does `any` capture that?

**Comment by b-strauss** [comment]
@rosendi What's the progress on this? Do you need assistance to complete this?

**Comment by guybedford** [comment]
I believe this is stalled pending completion. Help getting it to a completed point is always welcome.

**Comment by rosendi** [comment]
@guybedford @b-strauss I didn't finished it, because after a while I realized that the typings here might be useless, i.e. most of the time the system/builder config and bootstrap logic are a plain javascript. I don't close this issue, because maybe I'm wrong. @b-strauss have you got a practical example where the systemjs logic have to be a typescript file?

**Comment by b-strauss** [comment]
Well, how else would you call the API in a typesafe way? What if you want to dynamically load a library at runtime? The compiler will barf at every sight of `System.import`.

**Comment by aluanhaddad** [comment]
@guybedford I would be happy to help with this if there is still interest. I was actually writing my own version and was working on a PR for DefinitelyTyped when I stumbled upon this. While it's rare to use the SystemJS API directly from TypeScript, I have needed to do it a few times and it would certainly be nice to have.

**Comment by guybedford** [comment]
@aluanhaddad if you think it would still be useful, I'm happy to do what is best here.

**Comment by b-strauss** [comment]
> While it's rare to use the SystemJS API directly from TypeScript, I have needed to do it a few times and it would certainly be nice to have.

Angular 2 has async routing where the next view is loaded at runtime. This is not as uncommon as one might think.

**Comment by aluanhaddad** [comment]
@b-strauss that sounds like an excellent use case. 

@rosendi do you have any plans to resume work on this in the future?

**Pull request closed without merging**